### PR TITLE
[SPARK-25522][SQL] Improve type promotion for input arguments of elementAt function

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?
In ElementAt, when first argument is MapType, we should coerce the key type and the second argument based on findTightestCommonType. This is not happening currently. We may produce wrong output as we will incorrectly downcast the right hand side double expression to int.

```SQL
spark-sql> select element_at(map(1,"one", 2, "two"), 2.2);

two
```

Also, when the first argument is ArrayType, the second argument should be an integer type or a smaller integral type that can be safely casted to an integer type. Currently we may do an unsafe cast. In the following case, we should fail with an error as 2.2 is not a integer index. But instead we down cast it to int currently and return a result instead.

```SQL
spark-sql> select element_at(array(1,2), 1.24D);

1
```
This PR also supports implicit cast between two MapTypes. I have followed similar logic that exists today to do implicit casts between two array types.
## How was this patch tested?
Added new tests in DataFrameFunctionSuite, TypeCoercionSuite.

### CODE REVIEW

**Comment by viirya** [comment]
@dilipbiswal The two examples shown in the PR description are allowed in current and this proposes to disallow them? I'm a bit confused by the description.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -971,9 +971,36 @@ object TypeCoercion {
...
+          if !Cast.forceNullable(fromType, toType) =>
           implicitCast(fromType, toType).map(ArrayType(_, false)).orNull
 
+        // Implicit cast between Map types.
+        // Follows the same semantics of implicit casting between two array types.
+        // Refer to documentation above.
+        case (MapType(fromKeyType, fromValueType, fn), MapType(toKeyType, toValueType, true)) =>
+          val newFromType = implicitCast(fromKeyType, toKeyType).orNull
+          val newValueType = implicitCast(fromValueType, toValueType).orNull
+          if (newFromType != null && newValueType != null
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Don't we need to verify `!Cast.forceNullable(fromKeyType, toKeyType)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -2140,21 +2140,34 @@ case class ElementAt(left: Expression, right: Expression) extends GetMapValueUti
...
-        case _ => AnyDataType // no match for a wrong 'left' expression type
-      }
-    )
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, hasNull), e2: IntegralType) if (e2 != LongType) =>
+        Seq(ArrayType(e1, hasNull), IntegerType)
+      case (MapType(keyType, valueType, hasNull), e2) =>
+        TypeCoercion.findTightestCommonType(keyType, e2) match {
+          case Some(dt) => Seq(MapType(dt, valueType, hasNull), dt)
+          case _ => Seq.empty
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Previously it will do implicitly conversion on the right expr to map's key type. Map type is not converted at all.

This change finds the common type between right expr and map's key type and will do conversion on map and right expr.

Can you provide an example showing why this is needed?

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @viirya 
```select element_at(map(1,"one", 2, "two"), 2.2D);```
In this case, there is no key with value 2.2 in the map, right ? But we return "two" as the answer. If we simply cast the right side to the key type of the map, then we will cast 2.2 down to 2 and find "two".  But by finding a common type, we will compare both sides as double and correctly return "null" as the answer.

Here is presto's output : 
```SQL
presto:default> select element_at(MAP(array[1], array[1]), 1.1);
 _col0 
-------
 NULL  
(1 row)

Query 20180921_051929_00018_iif98, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto:default> select element_at(MAP(array[1], array[1]), 1.0);
 _col0 
-------
     1 
(1 row)
```

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Thanks @dilipbiswal for clarifying it. Can you update the PR description with such example, I feel it is more clear.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -971,9 +971,36 @@ object TypeCoercion {
         case (ArrayType(fromType, true), ArrayType(toType: DataType, false)) => null
 
         case (ArrayType(fromType, false), ArrayType(toType: DataType, false))
-            if !Cast.forceNullable(fromType, toType) =>
+          if !Cast.forceNullable(fromType, toType) =>
           implicitCast(fromType, toType).map(ArrayType(_, false)).orNull
 
+        // Implicit cast between Map types.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Do we have enough test for this new implicit cast rules?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 2 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: @viirya Thanks .. i have updated the description. Please let me know if it looks okay to you or anything is amiss.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 3 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @viirya actually i followed the arraytype example. Actually it has decent coverage, i think. If you can think of a case thats missing , i will add. I will also go through once again myself.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 4]
```diff
@@ -971,9 +971,36 @@ object TypeCoercion {
...
+        // Follows the same semantics of implicit casting between two array types.
+        // Refer to documentation above.
+        case (MapType(fromKeyType, fromValueType, fn), MapType(toKeyType, toValueType, true)) =>
+          val newFromType = implicitCast(fromKeyType, toKeyType).orNull
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @viirya Will change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 5]
```diff
@@ -971,9 +971,38 @@ object TypeCoercion {
         case (ArrayType(fromType, true), ArrayType(toType: DataType, false)) => null
 
         case (ArrayType(fromType, false), ArrayType(toType: DataType, false))
-            if !Cast.forceNullable(fromType, toType) =>
+          if !Cast.forceNullable(fromType, toType) =>
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit: revert this change.

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 6]
```diff
@@ -971,9 +971,38 @@ object TypeCoercion {
...
 
+        // Implicit cast between Map types.
+        // Follows the same semantics of implicit casting between two array types.
+        // Refer to documentation above.
+        case (MapType(fromKeyType, fromValueType, fn), MapType(toKeyType, toValueType, true))
+          if !Cast.forceNullable(fromKeyType, toKeyType) =>
+          val newKeyType = implicitCast(fromKeyType, toKeyType).orNull
+          val newValueType = implicitCast(fromValueType, toValueType).orNull
+          if (newKeyType != null && newValueType != null
+            && (!newKeyType.sameType(fromKeyType) || !newValueType.sameType(fromValueType))) {
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Do we need this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 7]
```diff
@@ -971,9 +971,38 @@ object TypeCoercion {
         case (ArrayType(fromType, true), ArrayType(toType: DataType, false)) => null
 
         case (ArrayType(fromType, false), ArrayType(toType: DataType, false))
-            if !Cast.forceNullable(fromType, toType) =>
+          if !Cast.forceNullable(fromType, toType) =>
           implicitCast(fromType, toType).map(ArrayType(_, false)).orNull
 
+        // Implicit cast between Map types.
+        // Follows the same semantics of implicit casting between two array types.
+        // Refer to documentation above.
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Could you add a note that we can't use force-nullable key types.

**Review by ueshin — COMMENTED** [review]
@dilipbiswal Seems like the examples in the description is the other way around?

**Comment by ueshin** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 6 above)
Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @ueshin U r right. Let me take it out and run the tests.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 8]
```diff
@@ -257,12 +257,48 @@ class TypeCoercionSuite extends AnalysisTest {
...
+    }
+    val sourceMapExprWithValueNull =
+      CreateMap(Seq(Literal.default(sourceType.keyType),
+        Literal.create(null, sourceType.valueType)))
+    targetNotNullableTypes.foreach { targetType =>
+      val x = 10
+      val castDefault =
+        TypeCoercion.ImplicitTypeCasts.implicitCast(sourceMapExprWithValueNull, targetType)
+      assert(castDefault.isEmpty,
+        s"Should not be able to cast $sourceType to $sourceType, but got $castDefault")
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: `to $sourceType` -> `to $targetType`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 9]
```diff
@@ -257,12 +257,48 @@ class TypeCoercionSuite extends AnalysisTest {
...
+    }
+
+    shouldNotCast(MapType(DoubleType, DoubleType, valueContainsNull = false),
+      CalendarIntervalType)
```
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: why there is a test for `CalendarIntervalType`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 10]
```diff
@@ -974,6 +974,33 @@ object TypeCoercion {
...
+            if (!(Cast.forceNullable(fromKeyType, toKeyType) ||
+              Cast.forceNullable(fromValueType, toValueType))) =>
+          val newKeyType = implicitCast(fromKeyType, toKeyType).orNull
+          val newValueType = implicitCast(fromValueType, toValueType).orNull
+          if (newKeyType != null && newValueType != null) {
+            MapType(newKeyType, newValueType, false)
+          } else {
+            null
+          }
+
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit:
```
case (MapType(fromKeyType, fromValueType, fromNullable), MapType(toKeyType, toValueType, toNullable))
    if !Cast.forceNullable(fromKeyType, toKeyType) && Cast.resolvableNullability(fromNullable, toNullable)
  ....
  if (Cast.forceNullable(fromValueType, toValueType) $$ !toNullable) {
    null
  } else {
    MapType(newKeyType, newValueType, toNullable)
  }
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: then it's just one case

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 11]
```diff
@@ -2140,21 +2140,34 @@ case class ElementAt(left: Expression, right: Expression) extends GetMapValueUti
...
   override def inputTypes: Seq[AbstractDataType] = {
-    Seq(TypeCollection(ArrayType, MapType),
-      left.dataType match {
-        case _: ArrayType => IntegerType
-        case _: MapType => mapKeyType
-        case _ => AnyDataType // no match for a wrong 'left' expression type
-      }
-    )
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, hasNull), e2: IntegralType) if (e2 != LongType) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit:
```
case (arr: ArrayType, e2: IntegralType) if ... =>
  Seq(arr, IntegerType)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 12]
```diff
@@ -974,6 +974,25 @@ object TypeCoercion {
             if !Cast.forceNullable(fromType, toType) =>
           implicitCast(fromType, toType).map(ArrayType(_, false)).orNull
 
+        // Implicit cast between Map types.
+        // Follows the same semantics of implicit casting between two array types.
+        // Refer to documentation above. Make sure that both key and values
+        // can not by null after the implicit cast operation by calling forceNullable
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit: I guess this is `can not be null`?

Comment (dilipbiswal) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @viirya Thank you.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 13]
```diff
@@ -2140,21 +2140,34 @@ case class ElementAt(left: Expression, right: Expression) extends GetMapValueUti
...
   }
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    super.checkInputDataTypes() match {
-      case f: TypeCheckResult.TypeCheckFailure => f
-      case TypeCheckResult.TypeCheckSuccess if left.dataType.isInstanceOf[MapType] =>
-        TypeUtils.checkForOrderingExpr(mapKeyType, s"function $prettyName")
-      case TypeCheckResult.TypeCheckSuccess => TypeCheckResult.TypeCheckSuccess
+    (left.dataType, right.dataType) match {
+      case (ArrayType(e1, _), e2) if e2 != IntegerType =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: `e1` is not used, this can be `case (_: ArrayType, e2) if ...`

**Comment by dilipbiswal** [comment]
@cloud-fan Are we targeting this for 2.4 ? If so, i was wondering if we should make a restricted fix that affects only this function by adding a case in `FunctionArgumentConversion` for 2.4 while keeping this fix for master. Pl. let me know what you think.

**Comment by cloud-fan** [comment]
I'm merging it to master, please send another PR to 2.4, without touching the `findTightestCommonType`. Thanks!

**Comment by dilipbiswal** [comment]
@cloud-fan @ueshin @viirya Thank you very much.

**Comment by cloud-fan** [comment]
After thinking more, I'm going it to merge it to 2.4 directly, because:
1. `ImplicitTypeCasts` handles array type but not map type, which looks a bug to me
2. I checked all the implementations of `ImplicitCastInputTypes`, `ElementAt` is the only one that sets a concrete map type as the input type. So this change only affects `ElementAt`
3. the test cases added in this PR iterate all the cases and have a very good converage.

**Comment by dilipbiswal** [comment]
@cloud-fan Sounds good Wenchen. Thanks for the detailed analysis.
